src/librbd/ImageWatcher.cc

```diff
@@ -1516,6 +1524,10 @@ void ImageWatcher<I>::handle_rewatch_complete(int r) {
   CephContext *cct = m_image_ctx.cct;
   ldout(cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
 
+  if (r == -EBLOCKLISTED) {
```

Comment:
Tying into the
> is there a reason this can't be done directly `handle_rewatch_complete`?

question above, at least at first sight, the most robust approach would seem to be to call `handle_peer_notification` here on any error.  Both `reacquire_lock` and `handle_payload` below assume that there is an active watch and that would be the case only when `r == 0`.

Reply:
> I didn't get why calling `handle_peer_notification` in `handle_rewatch_complete` is more robust. Can you please elaborate here?

The "more robust" bit was mostly referring to checking for any error rather than just `EBLOCKLISTED` in `handle_rewatch_complete`.  If you do that, you have to contend with the possibility that `m_watch_blocklisted` wouldn't be set in `handle_request_lock` and do something directly in `handle_rewatch_complete` to cover other errors.

Reply:
> The "more robust" bit was mostly referring to checking for any error rather than just `EBLOCKLISTED` in `handle_rewatch_complete`.

okay. 

> If you do that, you have to contend with the possibility that `m_watch_blocklisted` wouldn't be set in `handle_request_lock` and do something directly in `handle_rewatch_complete` to cover other errors.


If I'm following you, then you're suggesting the following changes to this PR,
```
diff --git a/src/librbd/ImageWatcher.cc b/src/librbd/ImageWatcher.cc
index 4bacc41ac8b..f704e369f1c 100644
--- a/src/librbd/ImageWatcher.cc
+++ b/src/librbd/ImageWatcher.cc
@@ -628,14 +628,6 @@ void ImageWatcher<I>::handle_request_lock(int r) {
     return;
   }
 
-  // kick-start ExclusiveLock state machine to detect client blocklisting
-  if (m_watch_blocklisted) {
-    lderr(m_image_ctx.cct) << this << " make lock detect client blocklisting"
-                           << dendl;
-    m_image_ctx.exclusive_lock->handle_peer_notification(0);
-    return;
-  }
-
   if (r == -ETIMEDOUT) {
     ldout(m_image_ctx.cct, 5) << this << " timed out requesting lock: retrying"
                               << dendl;
@@ -1524,7 +1516,10 @@ void ImageWatcher<I>::handle_rewatch_complete(int r) {
   CephContext *cct = m_image_ctx.cct;
   ldout(cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
 
-  if (r == -EBLOCKLISTED) {
+  if (r < 0) {
+    if (m_image_ctx.exclusive_lock != nullptr) {
+      m_image_ctx.exclusive_lock->handle_peer_notification(r);
+    }
     return;
   }



```

Reply:
Yes, roughly.  Things to improve:
- locking -- pretty sure `owner_lock` needs to be held there
- logging -- `lderr(m_image_ctx.cct) << this << " rewatch error: " << cpp_strerror(r) << dendl;`
- pass `0` to `handle_peer_notification` instead of `r` since we are just pretending the peer responded

Reply:
Done. Thanks!